sensor/common/compliance/nodescan_handler_impl.go

```diff
@@ -6,35 +6,51 @@ import (
 	"github.com/stackrox/rox/generated/storage"
 	"github.com/stackrox/rox/pkg/centralsensor"
 	"github.com/stackrox/rox/pkg/concurrency"
+	"github.com/stackrox/rox/pkg/sync"
 )
 
 type nodeScanHandlerImpl struct {
 	nodeScans <-chan *storage.NodeScanV2
-	toCentral chan *central.MsgFromSensor
+	toCentral <-chan *central.MsgFromSensor
+	// toCentralMux prevents the race condition between Start() [writer] and ResponsesC() [reader]
+	toCentralMux *sync.Mutex
 
-	stopC    concurrency.ErrorSignal
-	stoppedC concurrency.ErrorSignal
+	stopC concurrency.ErrorSignal
+	// runFinishedC is signaled when the goroutine inside of run() finishes
+	runFinishedC concurrency.ErrorSignal
+	startedC     concurrency.Signal
+}
+
+func (c *nodeScanHandlerImpl) Stopped() concurrency.ReadOnlyErrorSignal {
+	return &c.runFinishedC
 }
 
 func (c *nodeScanHandlerImpl) Capabilities() []centralsensor.SensorCapability {
 	return []centralsensor.SensorCapability{centralsensor.NodeScanningCap}
 }
 
 func (c *nodeScanHandlerImpl) ResponsesC() <-chan *central.MsgFromSensor {
+	c.toCentralMux.Lock()
+	defer c.toCentralMux.Unlock()
 	return c.toCentral
 }
 
 func (c *nodeScanHandlerImpl) Start() error {
-	// c.run closes chan toCentral on exit, so we do not want to close twice in case of a restart
+	c.toCentralMux.Lock()
```

Comment:
The more I read this and look at other Sensor components, the more I wonder if this may be overkill and all my questions about how to use this library were not very relevant. This seems to be correct, but I also wonder if it's worth worrying so much about this since we know how these methods are called (well at least now I know because I finally looked into it). Not saying to remove it but I wanted to see what you think

Reply:
+1 to not overkill. In my opinion, it will be sufficient to check whether a Start() was once attempted, and, if so, return an error. I don't think there's a need to make this component restartable provided that others don't take advantage of it.

Reply:
I know this is an overkill a bit, but I think this investment is worth it. We may need to make all Sensor components restartable, so I thought that this may be also a good way to have the race & goroutine leak tests green.